File: mscanner/fastscores/FeatureCounter.py
"""Calculates the number of occurrences of each feature for
all articles in Medline between two dates."""

from __future__ import division
import numpy as nx
from path import path
import struct

from mscanner import update
from mscanner.medline.FeatureStream import FeatureStream


class FeatureCounter:
    """Class for calculating feature counts in a subset of Medline.

    @ivar docstream: Path to file containing feature vectors for documents to
    score, in L{mscanner.medline.FeatureStream.FeatureStream} format.
    
    @ivar numdocs: Number of documents in the stream of feature vectors.
    
    @ivar numfeats: Number of distinct features in Medline (length of the 
    vector of feature counts).

    @ivar mindate: YYYYMMDD integer: documents must have this date or later
    (default 11110101)
    
    @ivar maxdate: YYYYMMDD integer: documents must have this date or earlier
    (default 33330303)

    @ivar exclude: PMIDs that are not allowed to appear in the results    
    """
    
    counter_path = path(__file__).dirname() / "_FeatureCounter"
    """Executable file for counting features in a file"""
    
    def __init__(self,
                 docstream,
                 numdocs,
                 numfeats,
                 mindate=None,
                 maxdate=None,
                 exclude=set(),
                 ):
        if mindate is None: mindate = 10110101
        if maxdate is None: maxdate = 30330303
        update(self, locals())


    def py_counts(s):
        """Simply iterate over the documents and count how
        many times each feature occurs in the specified range
        
        @return: Number of documents counted, and vector of feature counts."""
        featcounts = nx.zeros(s.numfeats, nx.int32)
        docs = FeatureStream(open(s.docstream, "rb"))
        ndocs = 0
        try:
            for docid, date, features in docs:
                if (date >= s.mindate and date <= s.maxdate 
                    and docid not in s.exclude):
                    featcounts[features] += 1
                    ndocs += 1
        finally:
            docs.close()
        return ndocs, featcounts


    def c_counts(s):
        """Pipes parameters to a C program that parses the stream
        of documents with features, which counts the number
        of occurrences of each feature, only considering documents
        added to Medline in the specified date range.
        
        @return: Number of documents counted, and vector of feature counts."""
        import subprocess as sp
        p = sp.Popen([
            s.counter_path, 
            s.docstream,
            str(s.numdocs),
            str(s.numfeats),
            str(s.mindate),
            str(s.maxdate),
            str(len(s.exclude)),
            ], stdout=sp.PIPE, stdin=sp.PIPE)
        p.stdin.write(nx.array(sorted(s.exclude)))
        # First integer of output is the number of documents parsed
        ndocs = struct.unpack("I", p.stdout.read(4))[0]
        # Then a vector of feature counts
        featcounts = nx.fromfile(p.stdout, nx.int32, s.numfeats)
        return ndocs, featcounts

File: mscanner/fastscores/_FeatureCounter.c
/* Calculate the number of occurrences of each feature within a 
specified date range.

Usage:

./featcounts \
[citations] \
[numdocs] \
[numfeats] \
[mindate] \
[maxdate] \
> feature_socres

  The [citations] file consists of [numcites] records, which are
  binary records, which can be expressed as the structure::

  struct {
    unsigned int pmid; // PubMed ID of citation
    unsigned int date; // Record completion date
    unsigned short nfeatures; // Number of features
    unsigned short features[nfeatures]; // Feature vector
  };

  The output is a list of [numfeats] integers representing the
  feature counts.
  
Copyright (C) 2007 Graham Poulter
*/

/* This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>. */

#include <stdio.h>
#include <stdlib.h>


// Search sorted array A of length N for needle.
// Return 1 if we find the needle, 0 if we do not
// http://en.wikipedia.org/wiki/Binary_search
int binary_search(int *A, int N, int needle) {
    int low = 0;
    int high = N-1;
    int mid = 0;
    while (low <= high) {
        mid = (low + high) / 2;
        if (A[mid] > needle) {
            high = mid - 1;
        } else if (A[mid] < needle) {
            low = mid + 1;
        } else {
            return 1;
        }
    }
    return 0;
}


int main (int argc, char **argv)
{
    // Parameters
    char *cite_filename = argv[1];    // Name of citations file
    int numcites = atoi (argv[2]);    // Number of citations
    int numfeats = atoi (argv[3]);    // Number of features
    int mindate = atoi (argv[4]);     // Minimum date to consider
    int maxdate = atoi (argv[5]);     // Maximum date to consider
    int numexcluded = atoi (argv[6]); // Number of excluded citations
    
    // Loop variables
    FILE *citefile = NULL; // File with citation scores
    int pi = 0; // Loop variable: number of PubMed ID's so far
    int fi = 0; // Loop variable: index into feature vector
    int date = 0; // Date of the current citation
    int pmid = 0; // PubMed ID of the current citation
    int ndocs = 0; // Number of documents counted
    unsigned short featvec_size = 0; // Size of current feature vector
    unsigned short featvec[1000]; // Maximum of 1000 features per citation

    // Allocate space for excluded PMIDs 
    int *excluded = (int*) malloc (numexcluded * sizeof(int));
    
    // Allocate space for feature counts 
    int *featcounts = (int*) malloc (numfeats * sizeof(int));

    // Read excluded PMIDs from input
    fread(excluded, sizeof(int), numexcluded, stdin);

    // Initialise feature counts to zero
    for(fi = 0; fi < numfeats; fi++) featcounts[fi] = 0;

    // Calculate citation scores
    citefile = fopen(cite_filename, "rb");
    for(pi = 0; pi < numcites; pi++) {
        // Read feature vector from the binary file
        fread(&pmid, sizeof(unsigned int), 1, citefile);
        fread(&date, sizeof(unsigned int), 1, citefile);
        fread(&featvec_size, sizeof(unsigned short), 1, citefile);
        fread(featvec, sizeof(unsigned short), featvec_size, citefile);
        // Don't bother if the date is outside the range
        if ((date < mindate) || (date > maxdate)) {
            continue;
        }
        // Don't bother if it is a member of exclude
        if (binary_search(excluded, numexcluded, pmid) == 1) {
            continue;
        }
        // Add up the feature scores
        for(fi = 0; fi < featvec_size; fi++)
            featcounts[featvec[fi]]++;
        // Counted features for one more document
        ndocs++;
    }
    fclose(citefile);

    // Print number of docs, feature counts before returning from main
    fwrite(&ndocs, sizeof(int), 1, stdout);
    fwrite(featcounts, sizeof(int), numfeats, stdout);
    return 0;
}
